routing/missioncontrol.go

```diff
@@ -287,7 +220,108 @@ func generateBandwidthHints(sourceNode *channeldb.LightningNode,
 // if no payment attempts have been made.
 func (m *MissionControl) ResetHistory() {
 	m.Lock()
-	m.failedEdges = make(map[EdgeLocator]time.Time)
-	m.failedVertexes = make(map[route.Vertex]time.Time)
+	m.history = make(map[route.Vertex]*nodeHistory)
 	m.Unlock()
 }
+
+// getEdgeProbability is expected to return the success probability of a payment
+// from fromNode along edge.
+func (m *MissionControl) getEdgeProbability(fromNode route.Vertex,
+	edge EdgeLocator) float64 {
+
+	m.Lock()
+	defer m.Unlock()
+
+	nodeHistory, ok := m.history[fromNode]
+	if !ok {
+		return 1
+	}
+
+	return m.getEdgeProbabilityForNode(nodeHistory, edge.ChannelID)
+}
+
+// getEdgeProbabilityForNode estimates the probability of successfully
+// traversing a channel based on the node history.
+func (m *MissionControl) getEdgeProbabilityForNode(nodeHistory *nodeHistory,
+	channelID uint64) float64 {
+
+	now := m.now()
+
+	// Calculate the last failure of the given edge. A node failure is
+	// considered a failure that would have affected every edge.
+	var lastFailure time.Time
+	if nodeHistory.lastFail != nil {
+		lastFailure = *nodeHistory.lastFail
+	}
+
+	lastEdgeFailure, ok := nodeHistory.channelLastFail[channelID]
+	if ok {
+		if lastEdgeFailure.After(lastFailure) {
+			lastFailure = lastEdgeFailure
+		}
+	}
+
+	// If we are still in the hard prune window, return probability 0.
```

Comment:
could we consider letting this happen, and instead rely on the payment timing out instead? That would allow long-lived path finding.

Reply:
Changed this part of the PR. Currently mission control isn't concerned with 'hard pruning' anymore. Instead, a minimum probability restriction is added to path finding. Setting this to a low value should help prevent trying unlikely routes in a short loop.

Reply:
For long payment timeouts it may actually be desired that we go over previously failed routes again to see if conditions improved.